### ISSUE TITLE
Оформление библиотеки, распределённые таблицы

### ISSUE THREAD

**Issue by iprst** [terratensor/regular-library#2]
Есть мысли о вариантах внутренней организации большой библиотеки.

Схема представления:

1. Вычисляем самые распространённые в полной библиотеке токены, общие для большинства книг. 
2. Вычисляем для каждой книги облако её наиболее частых токенизаций.
3. Удаляем из облака общие токены большинства, таким образом оставляя собственный «след» книги. 
4. Используем полученное тонкое облако в оформлении, показываем его рядом с параграфом. 
5. Между тонкими облаками можно установить отношения и расстояния.

Таким образом можно реализовать структурирование данных по вектору от параграфа к книге — найденные «тонкие облака» будут вполне чётко показывать направление мысли в книге, из которой взят параграф. 

Таким же образом можно разметить и другие характеристики, если, например, сделать не вычитание «общего облака» из «частного облака», а вместо этого сначала найти «специальное облако» например по интересующей терминологии, например набор концептуальных терминов, и смотреть, какие из этих терминов есть в книге, и показывать их вторым списком ниже первого облака. 

<details>
<summary>Условная картинка</summary>

![image310](https://github.com/terratensor/regular-library/assets/129882753/3e937cb6-4206-4c3b-a850-113d9a94268e)

</details>

**Comment by audetv** [comment]
У меня затык с пунктом 1 и 2, попробую сформировать, у меня почему-то получаются вложенные множества, сумма книг токенов будет всегда в общей сумме всех токенов. Они взаимоисключаются и получается идеальный мусорный след в каждой книге), т.к. мусорные токены будут самыми уникальными.  Но тут видимо, все дело в методе как определять «самые распространенные» токены. 
Надо как то соотношение определять, как мы это делали со средней длинной параграфа, что-то среднее квадратичное…

Мысль интересная, но у меня пока даже с оракулом, что то не получается представить цепочку до результата. Буду думать.

**Comment by iprst** [comment]
> У меня затык с пунктом 1 и 2, попробую сформировать, у меня почему-то получаются вложенные множества, сумма книг токенов будет всегда в общей сумме всех токенов. Они взаимоисключаются и получается идеальный мусорный след в каждой книге), т.к. мусорные токены будут самыми уникальными. Но тут видимо, все дело в методе как определять «самые распространенные» токены.

Моё представление такое — есть в приниципе общие слова, которые часто попадаются в книгах. Во всех в целом, или если бы мы били набор книг на разные таблицы — на каждый набор в целом. Мы имеем список самых частых слов, допустим это будет первая тысяча или первые десять тысяч. Получили п.1. Самые частые — это «мусор». 

Для каждой книги мы можем отдельно посчитать частотность? Или в рамках этой базы это невозможно? Если возможно отдельно посчитать частотность для каждой книги, считаем. Получаем её список частотных слов. Отмечаем в полученном списке все слова, которые не попали в общий мусор. Такие слова существуют, но находятся не на самом верху общего списка. 

> Надо как то соотношение определять, как мы это делали со средней длинной параграфа, что-то среднее квадратичное…

Очень сложно на глаз. Надо видеть хотя бы 2-3 таблицы — одну общую и пару частоток отдельных книг. 

> Мысль интересная, но у меня пока даже с оракулом, что то не получается представить цепочку до результата. Буду думать.

Думать надо лёжа. 
А набрасывать сюда сидя.
)))

**Comment by audetv** [comment]
> Для каждой книги мы можем отдельно посчитать частотность? Или в рамках этой базы это невозможно? Если возможно отдельно посчитать частотность для каждой книги, считаем. Получаем её список частотных слов. Отмечаем в полученном списке все слова, которые не попали в общий мусор. Такие слова существуют, но находятся не на самом верху общего списка.

Я не нашел как посчитать частотность для книги, команды kewords, squggest применяются для всей БД (таблицы),но сегодня пришла идея, что мы можем сделать это  так: дописать программу, создать дополнительную временную таблицу: library_temporarily_table, в нее записывать параграфы книги и вызывать нужные call keywords, получать цифры и сохранять их в БД, после удалять записи из таблицы, и делать запись уже следующей книги, если нужно, то можно выгрузить все словоформы для каждой книги и сохранить отдельно. Структуру таблиц надо подумать, пока набрасываю в тему. Запускать это можно или во время парсинга файлов или после отдельной утилитой. Надо подумать, как лучше, наверное в парсере было бы лучше, т.к. если распространять программу, конечный продукт, то одна утилита, которая сделает сразу все нужные действия с файлами выглядит лучше, чем сборник утилит для разных нужд. А так указал папку с файлами и пошел процесс индексации и необходимых расчетов. 

Т.е. в процессе обработки файлов пишем параграфы книги во временную таблицу, рассчитываем частотность для книги, сохраняем нужные данные для книги, таблицы и т.д. и записываем эту же книгу в общую БД librrary. 

Получить словоформы в книге можно перебором всех букв алфавита и цифр, примеры запросов, результаты будут отсортированы по параметру hits:

```
CALL KEYWORDS('а*', 'library', 1 as stats, 'hits' as sort_mode);
```
```
CALL KEYWORDS('б*', 'library', 1 as stats, 'hits' as sort_mode);
```
```
CALL KEYWORDS('1*', 'library', 1 as stats, 'hits' as sort_mode);
```

Без примеров результатов, очень большие списки получаются, проверить, посмотреть наглядно можно в консоли.

**Comment by iprst** [comment]
> дописать программу, создать дополнительную временную таблицу: library_temporarily_table, в нее записывать параграфы книги и вызывать нужные call keywords, получать цифры и сохранять их в БД, после удалять записи из таблицы, и делать запись уже следующей книги, если нужно, то можно выгрузить все словоформы для каждой книги и сохранить отдельно

Выглядит большим усложнением и тратой ресурсов. Возможно идею с частотой слов и облаком нужно пока оставить на проработку в фоновом режиме?

> наверное в парсере было бы лучше, т.к. если распространять программу, конечный продукт, то одна утилита, которая сделает сразу все нужные действия с файлами выглядит лучше, чем сборник утилит для разных нужд. А так указал папку с файлами и пошел процесс индексации и необходимых расчетов

Логично.

> Получить словоформы в книге можно перебором всех букв алфавита и цифр, примеры запросов, результаты будут отсортированы по параметру hits:

Как в командах ограничить показ по критерию hits только больше, например 1000?

**Comment by audetv** [comment]
> Выглядит большим усложнением и тратой ресурсов. Возможно идею с частотой слов и облаком нужно пока оставить на проработку в фоновом режиме?

Да, усложнение, но другого способа мне пока в голову не пришло. Пусть работает в фоне.

> Как в командах ограничить показ по критерию hits только больше, например 1000?

Не видел такой команды, посмотрю ещё, limit не работал, но тут надо добавить фильтр по критерию hits > 1000. скорее всего этого тоже нет, я вчера изучал документацию по этим командам и там довольно скудный список команд-параметров для keywords, я их все опробовал, возможно есть что-то не документированное. Позже найдем.

**Comment by iprst** [comment]
> Да, усложнение, но другого способа мне пока в голову не пришло. Пусть работает в фоне.

Надо сообразить, как это всё соединяется в прикладных задачах типа геоматрицы. Сейчас то с книгами вообще не проблема что-то найти из информации, почти любой «хитрый» (скорее натуральный) запрос в К150 даёт на порядок более полезную выдачу, чем любой поисковик, поскольку информация ограничена книгами. Куча неожиданно полезного вылазит просто из банальных казалось бы запросов, которые в обычном поисковике дадут только мусор. Поэтому сама полезность этой затеи уже очевидна, а как это будет оформлено не столь важно сейчас. Важнее понять, как это сопрягается с геоматрицей например. Может попробовать собрать базу для геоматрицы из сущностей с расчётом хэшей, и далее осознать, как две эти базы связаны? 

> Не видел такой команды, посмотрю ещё, limit не работал, но тут надо добавить фильтр по критерию hits > 1000. скорее всего этого тоже нет, я вчера изучал документацию по этим командам и там довольно скудный список команд-параметров для keywords, я их все опробовал, возможно есть что-то не документированное. Позже найдем.

В теории наверное я могу открыть внешнюю консоль (или настроить внутреннюю в докере) с достаточным количеством строк, чтобы результаты не обрезались. Чтобы посмотреть в таблице этого будет достаточно.

**Comment by audetv** [comment]
> Может попробовать собрать базу для геоматрицы из сущностей с расчётом хэшей, и далее осознать, как две эти базы связаны?

Да, хороший вариант и от этого уже далее планировать действия и усовершенствования программы.

> В теории наверное я могу открыть внешнюю консоль (или настроить внутреннюю в докере) с достаточным количеством строк, чтобы результаты не обрезались. Чтобы посмотреть в таблице этого будет достаточно.

Можно сделать так, открыть консоль cmd, перейти в папку на нужном диске, например `d:`, `cd terratensor`.
Далее набрать команду и указать вывод результатов не в консоль а в файл:
```
docker exec -it regular-library-manticore mysql -e "CALL KEYWORDS('а*', 'library', 1 as stats, 'hits' as sort_mode)" > keywords.txt
```
```
docker exec -it regular-library-manticore mysql -e "CALL KEYWORDS('й*', 'library', 1 as stats, 'hits' as sort_mode)" > keywords.txt
```

где `> keywords.txt` — имя файла, в который будет сохранен результат запроса.

Можно это и в докер консоли запустить, но там будет не удобно файлы смотреть, лучше cmd или любую другую.

Вот так это в докер консоли выглядит:
```
root@42222448678c:/var/lib/manticore# mysql -e "CALL KEYWORDS('а*', 'library', 1 as stats, 'hits' as sort_mode)" > keywords.txt
root@42222448678c:/var/lib/manticore# ls
binlog.001  binlog.002  binlog.lock  binlog.meta  data  keywords.txt  query.log  replication  searchd.log
root@42222448678c:/var/lib/manticore# 
```

<details><summary>Details</summary>
<p>

![2023-10-04_15-52-38](https://github.com/terratensor/regular-library/assets/10896447/20d49d16-4025-4334-8361-b9bf618c54e7)


</p>
</details>

**Comment by iprst** [comment]
> Можно сделать так, открыть консоль cmd, перейти в папку на нужном диске, например `d:`, `cd terratensor`.
> Далее набрать команду и указать вывод результатов не в консоль а в файл:

Всё понял, благодарю. 

> Да, хороший вариант и от этого уже далее планировать действия и усовершенствования программы.

Ага, и одновременно имея в виду новую К150 нужно посматривать в неё время от времени и сопоставлять как с гуглобусом, так и с базой геохэшей. Оно соберётся в систему «руками».

**Comment by iprst** [comment]
Эту мысль с облаком хочется подключить к обновлению V2
https://github.com/terratensor/library/issues/3

**Comment by audetv** [comment]
Можно почитать связанную математику, возможно будет полезно и по теме.
Закон Ципфа 
и LDA(latent Dirichlet allocation) распределение Дирихле

**Comment by iprst** [comment]
Собирался открыть дискуссию в духе «что изучить», в которой можно было бы набрасывать вопросы и ответы. [Открыл дискуссию](https://github.com/orgs/terratensor/discussions/18).
